File: leetcode/code/463.island_perimeter.go
package main

func islandPerimeter(grid [][]int) int {
    if grid == nil {
        return 0
    }

    var result int

    for ri, row := range grid {
        for ci, point := range row {
            if point == 1 {
                result += 4

                if ri > 0 && grid[ri - 1][ci] == 1 {
                    result -= 2
                }
                if ci > 0 && grid[ri][ci - 1] == 1 {
                    result -= 2
                }
            }
        }
    }

    return result
}


func main() {
    grid := [][]int{
        {0, 1, 0, 0},
        {1, 1, 1, 0},
        {0, 1, 0, 0},
        {1, 1, 0, 0},
    }
    if islandPerimeter(grid) != 16 {
        println("buggy")
    }
}

File: leetcode/code/463.island_perimeter.py
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        result = 0

        for ri, row in enumerate(grid):
            for ci, col in enumerate(row):
                if col == 1:
                    if ri == 0 or grid[ri - 1][ci] == 0:
                        result += 1
                    if ci == len(row) - 1 or grid[ri][ci + 1] == 0:
                        result += 1
                    if ri == len(grid) - 1 or grid[ri + 1][ci] == 0:
                        result += 1
                    if ci == 0 or grid[ri][ci - 1] == 0:
                        result += 1
        return result


if __name__ == "__main__":
    grid = [
        [0, 1, 0, 0],
        [1, 1, 1, 0],
        [0, 1, 0, 0],
        [1, 1, 0, 0],
    ]
    assert Solution().islandPerimeter(grid) == 16
